sentinel-dashboard/src/main/java/com/alibaba/csp/sentinel/dashboard/repository/metric/InMemoryMetricsRepository.java

```diff
@@ -46,46 +47,68 @@
      */
     private Map<String, Map<String, LinkedHashMap<Long, MetricEntity>>> allMetrics = new ConcurrentHashMap<>();
 
-    private final ReentrantReadWriteLock readWriteLock = new ReentrantReadWriteLock();
 
+    private Map<String, ReadWriteLock> appLockMap = new ConcurrentHashMap(); //lock seperate by app
 
-    @Override
-    public void save(MetricEntity entity) {
-        if (entity == null || StringUtil.isBlank(entity.getApp())) {
-            return;
-        }
-        readWriteLock.writeLock().lock();
-        try {
-            allMetrics.computeIfAbsent(entity.getApp(), e -> new HashMap<>(16))
-                    .computeIfAbsent(entity.getResource(), e -> new LinkedHashMap<Long, MetricEntity>() {
-                        @Override
-                        protected boolean removeEldestEntry(Entry<Long, MetricEntity> eldest) {
-                            // Metric older than {@link #MAX_METRIC_LIVE_TIME_MS} will be removed.
-                            return eldest.getKey() < TimeUtil.currentTimeMillis() - MAX_METRIC_LIVE_TIME_MS;
-                        }
-                    }).put(entity.getTimestamp().getTime(), entity);
-        } finally {
-            readWriteLock.writeLock().unlock();
-        }
 
+    @Override
+    public void save(MetricEntity entity, String app) {
+        allMetrics.computeIfAbsent(app, e -> new HashMap<>(16)).computeIfAbsent(entity.getResource(), e -> new LinkedHashMap<Long, MetricEntity>(100) {
+            @Override
+            protected boolean removeEldestEntry(Map.Entry<Long, MetricEntity> eldest) {
+                return eldest.getKey() < System.currentTimeMillis() - MAX_METRIC_LIVE_TIME_MS;
```

Comment:
`TimeUtil.currentTimeMillis()`  is a optimized version  for `System.currentTimeMillis()` , so here `TimeUtil.currentTimeMillis()` may be better